### PR TITLE
Eval parsing

### PR DESCRIPTION
**Pull request by jlopatec**
Fixed Eval parsing issue and added tests:

The following are common parsing needs that failed in the old approach. Hitting these a lot.

1) LABEL before explanation 
LABEL: incorrect EXPLANATION: The user is asking about the availability of 

2) Missing the label but just having the "rail" this is common customer experience 
incorrect EXPLANATION: The user is asking about the availability of

### CODE REVIEW

**Comment by mikeldking** [comment]
@sourcery-ai review

**Comment by anticorrelator** [comment]
I'm curious about when we're getting labels before explanations when we're using chain-of-prompt prompting

**Comment by jgomes168** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/phoenix-evals/src/phoenix/evals/templates.py [hunk 1]
```diff
@@ -186,12 +186,55 @@ def score(self, rail: str) -> float:
...
+    """
+    if not raw_string or not raw_string.strip():
+        return raw_string  # Fallback to the whole string if no label delimiter is found
+
+    # Sort rails by length (longest first) to avoid substring matching issues
+    sorted_labels = sorted(rails, key=len, reverse=True) if rails else None
+
+    # First, try to find a label after "LABEL:" keyword (case insensitive)
     label_delimiter = r"\W*label\W*"
     parts = re.split(label_delimiter, raw_string, maxsplit=1, flags=re.IGNORECASE)
```
Comment (mikeldking) [inline packages/phoenix-evals/src/phoenix/evals/templates.py]: I think here you could determine if the response is out of order @jgomes168

File: packages/phoenix-evals/tests/phoenix/evals/templates/test_template.py [hunk 2]
```diff
@@ -64,3 +65,232 @@ def test_template_with_alternate_delimiters():
...
+    # Multiple LABEL keywords - should use first one
+    assert (
+        parse_label_from_chain_of_thought_response(
+            "LABEL: incorrect EXPLANATION: something LABEL: correct", rails=rails
+        )
+        == "incorrect"
+    )
+
+
+def test_parse_label_from_chain_of_thought_response_handles_edge_cases():
```
Comment (mikeldking) [inline packages/phoenix-evals/tests/phoenix/evals/templates/test_template.py]: You can use pytest fixtures for this. As is it's sorta hard to read

File: packages/phoenix-evals/tests/phoenix/evals/templates/test_template.py [hunk 3]
```diff
@@ -64,3 +65,232 @@ def test_template_with_alternate_delimiters():
...
+    """Test edge cases to ensure complete coverage of all return paths."""
+    rails = ["relevant", "irrelevant"]
+
+    # Empty/whitespace input (return path 1)
+    assert parse_label_from_chain_of_thought_response("", rails=rails) == ""
+    assert parse_label_from_chain_of_thought_response("   ", rails=rails) == "   "
```
Comment (mikeldking) [inline packages/phoenix-evals/tests/phoenix/evals/templates/test_template.py]: extract into it's own test case. Also this test case frankly seems ridiculous. Why is this useful?

File: packages/phoenix-evals/src/phoenix/evals/templates.py [hunk 4]
```diff
@@ -186,12 +186,55 @@ def score(self, rail: str) -> float:
...
-    if len(parts) == 2:
-        return parts[1]
-    return raw_string  # Fallback to the whole string if no label delimiter is found
+
+    # Determine which text to extract the first word from
+    if len(parts) == 2 and (label_part := parts[1].strip()):
+        # Found "label" keyword with content after it
+        text_to_parse = label_part
```
Comment (mikeldking) [inline packages/phoenix-evals/src/phoenix/evals/templates.py]: this is not going to be true if things are out of order

**Review by mikeldking — CHANGES_REQUESTED** [review]
Blocking. This is ballooning into very different requirements.

Let's cleanup the code to handle out of order label parts and also similar "label" overlap problems. It's not clear from this code that it's trying to solve either of these well. 

It's also kinda impossible to follow.

File: packages/phoenix-evals/src/phoenix/evals/templates.py [hunk 5]
```diff
@@ -187,11 +187,24 @@ def score(self, rail: str) -> float:
...
-    label_delimiter = r"\W*label\W*"
-    parts = re.split(label_delimiter, raw_string, maxsplit=1, flags=re.IGNORECASE)
-    if len(parts) == 2:
-        return parts[1]
-    return raw_string  # Fallback to the whole string if no label delimiter is found
+    # Locate the first occurrence of "label"
+    match = re.search(r"\blabel\b", raw_string, flags=re.IGNORECASE)
+    if not match:
+        # No "label" keyword – return the input unchanged
+        return raw_string
```
Comment (mikeldking) [inline packages/phoenix-evals/src/phoenix/evals/templates.py]: this seems like a bad fallback. Need to add a test case for something like:

"correct Explanation - because it is so"

File: packages/phoenix-evals/src/phoenix/evals/templates.py [hunk 6]
```diff
@@ -187,11 +187,24 @@ def score(self, rail: str) -> float:
...
+
+    # Slice everything after the word "label"
+    remainder = raw_string[match.end() :].lstrip(" :.-\t")
```
Comment (mikeldking) [inline packages/phoenix-evals/src/phoenix/evals/templates.py]: what if the explanation comes first? This seems incorrect. What about

"because it is so Label: correct" or "Explanation: because it is so Label: correct"

File: packages/phoenix-evals/src/phoenix/evals/templates.py (hunk 5 above)
Comment (jgomes168) [reply packages/phoenix-evals/src/phoenix/evals/templates.py]: Not sure if you already chatted with Dustin, but we were planning to intentionally not capture this case (I'll defer to him for details)

File: packages/phoenix-evals/src/phoenix/evals/templates.py (hunk 6 above)
Comment (jgomes168) [reply packages/phoenix-evals/src/phoenix/evals/templates.py]: The chain of thought test should be captured by `test_llm_classify_cot`

**Comment by jgomes168** [comment]
@mikeldking lmk if you're good with the changes!

**Comment by mikeldking** [comment]
> @mikeldking lmk if you're good with the changes!

I defer to the reviewer here.

**Comment by jgomes168** [comment]
> > @mikeldking lmk if you're good with the changes!
> 
> I defer to the reviewer here.

Dustin approved, but I think your previous review is still blocking a merge
